File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function (root) {
    function solve(root, maxsum) {

        if (root == null) {
            return 0;
        }

        var leftsum = solve(root.left, maxsum);
        var rightsum = solve(root.right, maxsum);

        if (leftsum >= 0 && rightsum >= 0) {
            maxsum.value = Math.max(maxsum.value, root.val + leftsum + rightsum)
            return (root.val + Math.max(leftsum, rightsum))
        }
        else if (leftsum < 0 && rightsum < 0) {
            maxsum.value = Math.max(maxsum.value, root.val)
            return root.val

        }
        else{
            maxsum.value = Math.max(maxsum.value, root.val+ Math.max(leftsum, rightsum))
            return (root.val +  Math.max(leftsum, rightsum) )
        }
        
    }

    const maxsum = {
        value: -1001
    }

    solve(root, maxsum);

    return maxsum.value


};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int helper(int& maxsum, TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        int leftsum = helper(maxsum,root->left);
        int rightsum = helper(maxsum,root->right);
        if (leftsum < 0) {
            leftsum = 0;
        }
        if (rightsum < 0) {
            rightsum = 0;
        }
        int possiblemaxsum = leftsum + rightsum + root->val;
        maxsum = max(maxsum, possiblemaxsum);
        return (root->val + max(leftsum, rightsum));
    }
    int maxPathSum(TreeNode* root) {

        int maxsum = INT_MIN;

        int sum = helper(maxsum, root);

        return maxsum;
    }
};